### DISCUSSION TITLE
Add inverter

### DISCUSSION THREAD

**Discussion by timbms** [pvlib/pvlib-python#1710] [General]
[c0nr3f](https://github.com/c0nr3f) was asking in #1709 how to add inverters not (yet) included in database.
Like  Kostal Plenticore Plus. Happy to support because I precisely configured it for this inverter.

**Comment by cwhanse** [comment]
Thanks @timbms for transferring to a discussion. The original question:

"How can I add an inverter, that is neither present in the database nor on the nhrel website ? I use a Kostal Plenticore plus and besides the typical values like the voltages and max power and current, I don't know where to get C-values from. Any suggestions or some kind of tool that can be used to calculate these?"

@c0nr3f

**Reply by c0nr3f** [reply]
Thank you @timbms and @cwhanse for transferring the question!

**Comment by c0nr3f** [comment]
So how do I add this inverter? Is there a tool or did found the values in the datasheet? I use a Kostal Plenticore plus 5.5.

**Comment by cwhanse** [comment]
If you are willing to digitize the Pac vs. efficiency curves from the datasheet (and convert from efficiency back to Pdc input), you can use this function to get the parameters for the Sandia model.

https://github.com/pvlib/pvlib-python/blob/7a2ec9b4765124463bf0ddd0a49dcfedc4cbcad7/pvlib/inverter.py#L440

**Comment by c0nr3f** [comment]
@cwhanse Thanks, I will try that!

@timbms Can you share your parameters for your inverter model?

**Reply by timbms** [reply]
I will share as soon as I have access to my parameters

**Reply by timbms** [reply]
I proceeded exactly as described by @cwhanse by digitising the inverter's data sheet.
Attached you will find my Apple Numbers and an XLS export. [InverterDigitalization.zip](https://github.com/pvlib/pvlib-python/files/11161646/InverterDigitalization.zip). Export as CSV for usage below

```
# Generation of inverter parameters for Kostal Plenticore 7.0
# https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.inverter.fit_sandia.html
# Method as seen in test for pvlib
dc_voltage=[290.0,570.0,720.0]
dc_voltage_level=[0.1,0.2,0.3,0.5,0.75,1.0]
p_ac_0=7500
p_nt=7.6 # night time consumption
curves = pd.read_csv(r'InverterData.csv', sep= ';')
dc_power = curves['ac_power'] / curves['efficiency']
inverter_parameters = inverter.fit_sandia(ac_power=curves['ac_power'],
                                 dc_power=dc_power,
                                 dc_voltage=curves['dc_voltage'],
                                 dc_voltage_level=curves['dc_voltage_level'],
                                 p_ac_0=p_ac_0, p_nt=p_nt)
print(inverter_parameters)
system_two_arrays = pvsystem.PVSystem(arrays=[array_one, array_two],
                             inverter_parameters=inverter_parameters)
mc = modelchain.ModelChain(system_two_arrays, loca, aoi_model='no_loss',
                spectral_model='no_loss')
```
I haven't understood how pvlib handles the three independent strings (of which I use 2) with their independent mpp tracking. Happy to get your input.

**Reply by cwhanse** [reply]
When a PVSystem has more than one array, ModelChain directs (through PVsystem.get_ac) the DC outputs of the arrays to one of the `inverter.*_multi` functions (* is either `sandia` or `pvwatts`). These functions implement the multi-MPPT versions of the Sandia and PVWatts's DC-AC models.

The reference behind the `sandia_multi` function justifies fitting the inverter model parameters to the overall efficiency curves found on a typical inverter datasheet, even for multi-MPPT inverters.

Your remark actually points out an opportunity to improve the documentation for that function. It does not require the number of MPPT inputs as an inverter parameter, but doesn't tell you that. If I was a user, I'd wonder how it knows there are 3 MPPTs (one unused), and if somehow efficiency is being modeled incorrectly.

**Reply by c0nr3f** [reply]
Are we looking on different datasheets? How did you get the data from these unprecise diagrams off the datasheet? 
Did you take the fractionalized power and voltage levels out of real data or did ask kostal for that?
When running your code as is, it won't run since 'InverterData.csv' doesn't exist without manually changing your xlsx. I know, how to do this but I still don't know where you have your values from. The datasheet I have just hast some basic data, mostly the voltage ranges.
Also, .numbers just run on apple systems, do you have some kind of open source format?

**Reply by timbms** [reply]
I took the data sheets for the Plenticore from [Kostal's website](https://www.kostal-solar-electric.com/de-de/produkte/hybrid-wechselrichter/plenticore-plus/-/media/document-library-folder---kse/2020/12/15/13/06/db_plenticore-plus_en.pdf) and read the values from the graph. 
<img width="489" alt="Screenshot 2023-04-23 at 12 16 23" src="https://user-images.githubusercontent.com/5411131/233834892-6202e7b6-d006-410d-ad3b-f1465a58c1d5.png">
I put the data into Apple Numbers and exported to csv.

**Reply by c0nr3f** [reply]
Thanks! That was the final information I needed. Here the data for anybody with a Kostal Plenticore Plus 5.5:

fraction_of_rated_power;dc_voltage_level;ac_power;dc_voltage;efficiency;power ac max
0,1;Vmin;550;225;0,904;5500
0,2;Vmin;1100;225;0,940;5500
0,3;Vmin;1650;225;0,951;5500
0,5;Vmin;2750;225;0,961;5500
0,75;Vmin;4125;225;0,965;5500
1;Vmin;5500;225;0,965;5500
0,1;Vnom;550;570;0,928;5500
0,2;Vnom;1100;570;0,953;5500
0,3;Vnom;1650;570;0,961;5500
0,5;Vnom;2750;570;0,967;5500
0,75;Vnom;4125;570;0,971;5500
1;Vnom;5500;570;0,970;5500
0,1;Vmax;550;720;0,919;5500
0,2;Vmax;1100;720;0,945;5500
0,3;Vmax;1650;720;0,953;5500
0,5;Vmax;2750;720;0,959;5500
0,75;Vmax;4125;720;0,961;5500
1;Vmax;5500;720;0,961;5500

**Reply by timbms** [reply]
Great! Just a question to the maintainers. What is the process for sharing new inverter/module data?

**Reply by cwhanse** [reply]
> What is the process for sharing new inverter/module data?

Short answer: we don't have a process. But we can make one.

We could add the code and data that generates the inverter/module parameters to the Example Gallery. Having this example would be very helpful. 

However, adding Examples doesn't scale to become an archive of parameters for devices not available via the SAM files. To open a door to accept new parameter sets, my thoughts:
1) we probably don't want to add rows to the SAM files. They would be lost when we import a later version of the CEC files, and it confuses the source of these new rows.
2) we could create additional CSV files with the same structure as the SAM files, e.g., `pvlib\data\pvlib-inverters-2023.csv`. I would want to also keep the code that generated the new parameters. So that code would need to go into a new folder somewhere, but not as part of the pvlib code base. It feels like this option would have a number of complications that I haven't anticipated.
3) we could broaden the intended scope of pvlib/pvmodules to both modules and inverters, and archive the parameter generation code and its output there. I don't know how we could automate transfer of the parameters to pvlib-python, or if we'd want to.

**Reply by wholmgren** [reply]
I don't think pvlib should get into the business of distributing user-contributed module/inverter parameters. I'd recommend @mikofski's [pvfree](https://pvfree.azurewebsites.net/)

**Reply by adriesse** [reply]
Does pvfree support adding products?

**Reply by wholmgren** [reply]
My understanding is that currently @mikofski runs some kind of upload script or uses a django adminstration panel. But I expect that pvfree could be extended to support a user-contributed database without too much work. Perhaps users could contribute json or csv files via pull request, then github actions could run a script that uploads the new data to the database. I don't really know the infrastructure though. See https://github.com/pvlib/pvlib-python/issues/1345#issuecomment-982075372 for some previous discussion.

**Reply by mikofski** [reply]
Hi Folks, sorry I really thought I had responded somewhere, but maybe I forgot to click send. Anyway, yes we can add new inverters that fit the CEC model to pvfree, but first I'd like to explore and define how that should be structured and find a way to perhaps share some of the responsibility. Currently all data in pvfree is sourced exclusively from SAM, and I bear the entire burden of maintaining the site. Currently the "Source" key is parsed from the "Name" of the module because that's the way SAM CSV files are tabulated, so there's no column that explicitly states where the data came from.

**Reply by adriesse** [reply]
Alternatively this could be a commercial or semi-commercial undertaking as it requires some investment of time and effort.

**Reply by wholmgren** [reply]
A first step could be to brainstorm a repository/data structure and GitHub workflow that would allow for the standard data sources and user contributions. Then some pvlib examples of using that data directly from GitHub. Finally integrate the data into an application (pvfree or others).

**Reply by cwhanse** [reply]
> A first step could be to brainstorm a repository/data structure and GitHub workflow that would allow for the standard data sources and user contributions. 

That's half of the original motivation of pvlib/pvmodules. The other motivation is to make public the code that processes the CEC data to get the SAM files.

**Reply by adriesse** [reply]
> > A first step could be to brainstorm a repository/data structure and GitHub workflow that would allow for the standard data sources and user contributions.
> 
> That's half of the original motivation of pvlib/pvmodules. The other motivation is to make public the code that processes the CEC data to get the SAM files.

That's sounds pretty easy and not dependent on the rest.
